pkg/sync/grpc/grpc_sync.go

```diff
@@ -0,0 +1,186 @@
+package grpc
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"strings"
+	"time"
+
+	"google.golang.org/grpc/credentials/insecure"
+
+	"buf.build/gen/go/open-feature/flagd/grpc/go/sync/v1/syncv1grpc"
+	v1 "buf.build/gen/go/open-feature/flagd/protocolbuffers/go/sync/v1"
+
+	"github.com/open-feature/flagd/pkg/logger"
+	"github.com/open-feature/flagd/pkg/sync"
+	"google.golang.org/grpc"
+)
+
+const (
+	// Prefix for GRPC URL inputs. GRPC does not define a prefix through standard. This prefix helps to differentiate
+	// remote URLs for REST APIs (i.e - HTTP) from GRPC endpoints.
+	Prefix = "grpc://"
+
+	// Connection retry constants
+	// Back off period is calculated with backOffBase ^ #retry-iteration. However, when #retry-iteration count reach
+	// backOffLimit, retry delay fallback to constantBackOffDelay
+	backOffLimit         = 3
+	backOffBase          = 4
+	constantBackOffDelay = 60
+)
+
+type Sync struct {
+	Target     string
+	ProviderID string
+	Logger     *logger.Logger
+}
+
+func (g *Sync) Sync(ctx context.Context, dataSync chan<- sync.DataSync) error {
+	options := []grpc.DialOption{
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+	}
+
+	// initial dial and connection. Failure here must result in a startup failure
+	dial, err := grpc.DialContext(ctx, g.Target, options...)
+	if err != nil {
+		g.Logger.Error(fmt.Sprintf("Error establishing grpc connection: %s", err.Error()))
+		return err
```

Comment:
If the errors are being handled (logged) then is it necessary to return them too? Could this lead the errors to be logged multiple times?

Reply:
Sync providers are backed by an error group. And the caller (the `Runtime`) has the option to log or ignore logging if there's an error. In my view, logging and returning error gives better details when something goes wrong. Consider the following log for a startup failure, 

```
2023-02-14T08:48:18.905-0800    info    service/connect_service.go:110  metrics listening at 8014       {"component": "service"}
2023-02-14T08:48:18.908-0800    error   grpc/grpc_sync.go:54    error calling streaming operation: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing: dial tcp [::1]:9090: connect: connection refused"     {"component": "sync", "sync": "grpc"}
github.com/open-feature/flagd/pkg/sync/grpc.(*Sync).Sync
        /Users/kavindu.dodanduwa/Development/Git/flagd/pkg/sync/grpc/grpc_sync.go:54
github.com/open-feature/flagd/pkg/runtime.(*Runtime).Start.func2
        /Users/kavindu.dodanduwa/Development/Git/flagd/pkg/runtime/runtime.go:77
golang.org/x/sync/errgroup.(*Group).Go.func1
        /Users/kavindu.dodanduwa/go/pkg/mod/golang.org/x/sync@v0.1.0/errgroup/errgroup.go:75
2023-02-14T08:48:18.909-0800    fatal   cmd/start.go:122        rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing: dial tcp [::1]:9090: connect: connection refused"        {"component": "start"}
github.com/open-feature/flagd/cmd.glob..func1
        /Users/kavindu.dodanduwa/Development/Git/flagd/cmd/start.go:122
github.com/spf13/cobra.(*Command).execute
        /Users/kavindu.dodanduwa/go/pkg/mod/github.com/spf13/cobra@v1.6.1/command.go:920
github.com/spf13/cobra.(*Command).ExecuteC
        /Users/kavindu.dodanduwa/go/pkg/mod/github.com/spf13/cobra@v1.6.1/command.go:1044
github.com/spf13/cobra.(*Command).Execute
        /Users/kavindu.dodanduwa/go/pkg/mod/github.com/spf13/cobra@v1.6.1/command.go:968
github.com/open-feature/flagd/cmd.Execute
        /Users/kavindu.dodanduwa/Development/Git/flagd/cmd/root.go:37
main.main
        /Users/kavindu.dodanduwa/Development/Git/flagd/main.go:30
runtime.main
        /usr/local/go/src/runtime/proc.go:250
```

Reply:
I think in _legitimate_ error scenarios (ones that are unexpected or occur when you're first fitting things together) some duplicated errors is probably better than missing context.